src/coreclr/vm/ceemain.cpp

```diff
@@ -839,33 +839,13 @@ void EEStartupHelper()
 
 #ifdef PROFILING_SUPPORTED
         // Initialize the profiling services.
+        // THis must happen before the finalizer thread is stopped on its first wait.
```

Comment:
I do think this describes the required invariant accurately.

My concern is about making sure that the profilers receive [these ThreadCreated notification](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/threads.cpp#L749-L764) for the finalize or any other thread reliably. I think it means that `ProfilingAPIUtility::InitializeProfiling` is called before `SetupThread` is on any thread.

Reply:
I am wondering whether it would make sense to start the profiler thread in a different place on Windows vs. non-Windows to make both cases about as optimal as possible.

Reply:
Start it earlier on Windows since we know we will need to wait for it and it won't run away to do other things.

Start it later on Windows once there is no problem with the finalizer thread running away to do other things.

Reply:
My main concern is that `InitializeProfiling` would not try to set up a thread, directly or indirectly. It looks like it does not.

Reply:
InitializeProfiling looks fine with the current version of the change (except for the comment that I left feedback for). 

My suggestion is:
- Call `FinalizerThread::FinalizerThreadCreate` as one of the things in `EEStartupHelper` on Windows. It will maximize the chances that the finalizer thread is going to finish initializing COM by the time we need to wait for it.
- Call `FinalizerThread::FinalizerThreadCreate` as late as necessary on non-Windows to avoid stepping on the subtle landmines like debugger or profiler.

Reply:
I have realized that we would have to wait on the finalizer thread before calling HasStarted for this suggestion to work. You can ignore the suggestion. I assume that we will want to push this change to servicing so this should be a minimal fix.

The problem with debugger initialization that I have mentioned in the other comment may be real.

Reply:
Actually, we may need to introduce a wait like that on the finalizer thread to fix the debugger initialization problem.

Reply:
I think next to what GC does (reserves tons of VM, creates a few threads,..), it may be ok. 
But maybe one day GC will do all that in pay-for-play fashion. Fundamentally, it does not have to do it upfront.

The alternative is to start finalizer thread later on non-windows - as you've suggested. After the debugger is fully initialized.

Reply:
> Actually, we may need to introduce a wait like that on the finalizer thread to fix the debugger initialization problem.

Yes, to be sure that finalizer thread on Windows waits for debugger init we need to make it wait for EE thread to tell it what it can to proceed to `HasStarted()`

I've added that in the last change. Also in non-Windows case simply moved `FinalizerThreadCreate()` to a place after `g_pDebugInterface->StartupPhase2`